### PR TITLE
Shard Search Scroll failures consistency

### LINKED ISSUE #62046: SqlClearCursorActionIT.testAutoCursorCleanup failure
**Build scan**:
https://gradle-enterprise.elastic.co/s/vww2nhtqvllnq

**Repro line**:
```
 ./gradlew ':x-pack:plugin:sql:internalClusterTest' --tests "org.elasticsearch.xpack.sql.action.SqlClearCursorActionIT.testAutoCursorCleanup" \
  -Dtests.seed=ED48C0414093BFA8 \
  -Dtests.security.manager=true \
  -Dtests.locale=de \
  -Dtests.timezone=America/Campo_Grande \
  -Druntime.java=11
```

**Reproduces locally?**:
No

**Applicable branches**:
master

**Failure history**:
First ever failure I could find at least.

I have not muted the test because this is the first failure and it does not reproduce 

**Failure excerpt**:
```
java.lang.AssertionError: expected:<0> but was:<1>
	at __randomizedtesting.SeedInfo.seed([ED48C0414093BFA8:2A4953187A7F7CBA]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.elasticsearch.xpack.sql.action.SqlClearCursorActionIT.testAutoCursorCleanup(SqlClearCursorActionIT.java:85)
```

### LINKED ISSUE #62056: JdbcSqlSpecIT test {subselect.testSimpleSubSelect} failure
**Build scan**:
https://gradle-enterprise.elastic.co/s/2zcmgk52qs2uc

**Repro line**:
```
./gradlew ':x-pack:plugin:sql:qa:server:single-node:integTest' --tests "org.elasticsearch.xpack.sql.qa.single_node.JdbcSqlSpecIT" \
  -Dtests.method="test {subselect.testSimpleSubSelect}" \
  -Dtests.seed=32C4B0F4FD776FF6 \
  -Dtests.security.manager=true \
  -Dtests.locale=ar-YE \
  -Dtests.timezone=Africa/Luanda \
  -Druntime.java=11
```

**Reproduces locally?**:
No

**Applicable branches**:
master

**Failure history**:
This is the first failure of the test I could find. Leaving unmuted as not reproducible and is a one-off for now

**Failure excerpt**:
```
java.lang.AssertionError: test_emp should have no search contexts expected:<0> but was:<1>
	at __randomizedtesting.SeedInfo.seed([32C4B0F4FD776FF6:BA908F2E538B020E]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.elasticsearch.xpack.sql.qa.rest.RestSqlTestCase.assertNoSearchContexts(RestSqlTestCase.java:1096)
	at org.elasticsearch.xpack.sql.qa.jdbc.JdbcIntegrationTestCase.checkSearchContent(JdbcIntegrationTestCase.java:38)
```

### PR DESCRIPTION
**Pull request by jimczi**
Today some uncaught shard failures such as RejectedExecutionException skips the release of shard context
and let subsequent scroll requests access the same shard context again. Depending on how the other shards advanced,
this behavior can lead to missing data since scrolls always move forward.
In order to avoid hidden data loss, this commit ensures that we always release the context of shard search scroll requests whenever a failure occurs locally. 
The shard search context will no longer exist in subsequent scroll requests which will lead to consistent shard failures in the responses.

This change also modifies the retry tests of the reindex feature. Reindex retries scroll search request that contains a shard failure and move on whenever the failure disappears. That is not compatible with how scrolls work and can lead to missing data as explained above.
That means that reindex will now report scroll failures when search rejection happen during the operation instead of skipping document silently.

Finally this change removes an old TODO that was fulfilled with #61062.

Relates #61062

Note for reviewer: The last [commit](https://github.com/elastic/elasticsearch/pull/62061/commits/1f12505eac0e18d8077101a807a8d0c751d0700f) contains a fix for #62046 and #62056 since these tests failed in CI when checking this PR. The change ensures that we create a single searcher for scrolls.

Closes #62046
Closes #62056

### CODE REVIEW

**Review by nik9000 — COMMENTED** [review]
The reindex tests look right to me. I don't know the rest of the code well enough to be sure about it but I think that is most for @dnhatn .

**Review by dnhatn — APPROVED** [review]
LGTM with a small comment

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -643,13 +611,18 @@ private ReaderContext findReaderContext(ShardSearchContextId id, TransportReques
         return reader;
     }
 
+    private Releasable updateKeepAliveAndMarkAsUsed(ReaderContext reader, long keepAlive) {
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: ++, I pushed https://github.com/elastic/elasticsearch/pull/62061/commits/7d927699905a354353f242505aedf805019bade8

**Comment by dnhatn** [comment]
I've backported this in https://github.com/elastic/elasticsearch/commit/3fc35aa76e6cb2adda6879c335d805f6384923af.

**Pull request merged**